src/src/com/microsoft/aad/adal/AuthenticationContext.java

```diff
@@ -1430,23 +1431,31 @@ private AuthenticationResult localFlow(CallbackHandler callbackHandle,
 
         // Trying to find refresh token first, if existed, will try to use the refresh token. 
         Logger.v(TAG, "Checking refresh tokens");
-        final boolean useMultiResourceRefreshToken = cachedItem == null || cachedItem.getIsMultiResourceRefreshToken();
-        final RefreshItem refreshItem = getRefreshToken(request, useMultiResourceRefreshToken);
+        final RefreshItem refreshItem = getRefreshToken(request);
         AuthenticationResult authResult = null;
         if (!promptUser(request.getPrompt()) && refreshItem != null
                 && !StringExtensions.IsNullOrBlank(refreshItem.mRefreshToken)) {
-            Logger.v(TAG, "Refresh token is available and it will attempt to refresh token");
+            Logger.v(TAG, "Refresh token is available and will attempt to used it for a new access token.");
             try {
                 authResult = getTokenWithRefreshTokenAndUpdateCache(request, refreshItem);
             } catch (AuthenticationException authenticationException) {
-                ClientAnalytics.logEvent(
-                        InstrumentationIDs.REFRESH_TOKEN_REQUEST_FAILED,
-                        new InstrumentationPropertiesBuilder(request, authenticationException).build());
+                addLogEventForRequestFailureWithCause(InstrumentationIDs.REFRESH_TOKEN_REQUEST_FAILED, request, authenticationException);
                 callbackHandle.onError(authenticationException);
                 return null;
             }
             
-            if (authResult != null && !StringExtensions.IsNullOrBlank(authResult.getAccessToken())) {
+            // retry with MRRT if the refresh token request was done with FRT
+            if (!isAccessTokenReturned(authResult) && refreshItem != null) {
+                try {
+                    authResult = fallbackToMRRT(request, refreshItem);
+                } catch (final AuthenticationException authenticationException) {
+                    addLogEventForRequestFailureWithCause(InstrumentationIDs.REFRESH_TOKEN_REQUEST_FAILED, request, authenticationException);
+                    callbackHandle.onError(authenticationException);
+                    return null;
```

Comment:
I'm not going to cover this in this PR, I can submit separate PR for removing the early return in this method.